REGRESSION(285368@main): [iOS17] 2 TestWebKitAPI tests are constant crash (failure in EWS)

sysrqb opened the pull request:
#### 00e9cca4cf48e35e4d8ebe0774730b32943a03dd
<pre>
REGRESSION(285368@main): [iOS17] 2 TestWebKitAPI tests are constant crash (failure in EWS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=281770">https://bugs.webkit.org/show_bug.cgi?id=281770</a>
<a href="https://rdar.apple.com/138203988">rdar://138203988</a>

Reviewed by Wenson Hsieh.

We need to be careful about accessing VM state while the VM is being
initialized. Our telemetry collection now creates a VM where it wasn&apos;t needed
before. This patch avoids that issue by aborting telemetry collection if we
don&apos;t already have a vm.

Covered by existing crashing tests.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::vmIfExists):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptExecutionTelemetry):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::vmIfExists):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:

Canonical link: <a href="https://commits.webkit.org/285495@main">https://commits.webkit.org/285495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f80e7da13593601bcea93206edc8989ba521e977

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76839 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23883 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57151 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15659 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47107 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37579 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43757 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20152 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65613 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78512 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16896 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19502 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65599 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62692 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64871 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13174 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6954 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11191 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47873 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2660 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48940 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48685 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -920,7 +920,11 @@ RefCountedSerialFunctionDispatcher& ScriptExecutionContext::nativePromiseDispatc
 bool ScriptExecutionContext::requiresScriptExecutionTelemetry(ScriptTelemetryCategory category)
 {
-    Ref vm = this->vm();
+    JSC::VM* vmOrNull = this->existingVMOrNull();
```
```suggestion
    RefPtr vmOrNull = existingVMOrNull();
```

whsieh commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -270,6 +270,7 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
 
     virtual JSC::VM& vm() = 0;
     virtual Ref<JSC::VM> protectedVM();
+    virtual JSC::VM* existingVMOrNull() = 0;
```
I think the established convention for "getter that returns nullable reference to object that may not exist" would be `vmIfExists()`:

```
virtual JSC::VM* vmIfExists() const = 0;
```

(This should also be made `const` to reflect that it shouldn't instantiate a VM as a result of calling it)

Commit:
REGRESSION(285368@main): [iOS17] 2 TestWebKitAPI tests are constant crash (failure in EWS)
https://bugs.webkit.org/show_bug.cgi?id=281770
rdar://138203988

Reviewed by Wenson Hsieh.

We need to be careful about accessing VM state while the VM is being
initialized. Our telemetry collection now creates a VM where it wasn't needed
before. This patch avoids that issue by aborting telemetry collection if we
don't already have a vm.

Covered by existing crashing tests.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::vmIfExists):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptExecutionTelemetry):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::vmIfExists):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:

Canonical link: https://commits.webkit.org/285495@main
